javaslang/src/main/java/javaslang/collection/Collections.java

```diff
@@ -85,75 +108,62 @@
         });
     }
 
-    static Option<Integer> indexOption(int index) {
-        return Option.when(index >= 0, index);
+    @SuppressWarnings("unchecked")
+    static <C extends Traversable<T>, T> C removeAll(C source, Iterable<? extends T> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        final Set<T> removed = HashSet.ofAll(elements);
+        return (C) source.filter(e -> !removed.contains(e));
     }
 
-    // checks, if the *elements* of the given iterables are equal
-    static boolean areEqual(Iterable<?> iterable1, Iterable<?> iterable2) {
-        final java.util.Iterator<?> iter1 = iterable1.iterator();
-        final java.util.Iterator<?> iter2 = iterable2.iterator();
-        while (iter1.hasNext() && iter2.hasNext()) {
-            if (!Objects.equals(iter1.next(), iter2.next())) {
-                return false;
-            }
-        }
-        return iter1.hasNext() == iter2.hasNext();
+    @SuppressWarnings("unchecked")
+    static <C extends Traversable<T>, T> C removeAll(C source, Predicate<? super T> predicate) {
+        Objects.requireNonNull(predicate, "predicate is null");
+        return (C) source.filter(predicate.negate());
     }
 
-    // hashes the elements of an iterable
-    static int hash(Iterable<?> iterable) {
-        int hashCode = 1;
-        for (Object o : iterable) {
-            hashCode = 31 * hashCode + Objects.hashCode(o);
-        }
-        return hashCode;
+    static <C extends Traversable<T>, T> C removeAll(C source, T element) {
+        Objects.requireNonNull(element, "element is null");
+        return removeAll(source, List.of(element));
     }
 
-    static <T, U, R extends Traversable<U>> R scanLeft(Traversable<? extends T> traversable,
-                                                       U zero, BiFunction<? super U, ? super T, ? extends U> operation, Function<Iterator<U>, R> finisher) {
-
-        final Iterator<U> iterator = traversable.iterator().scanLeft(zero, operation);
-        return finisher.apply(iterator);
+    @SuppressWarnings("unchecked")
+    static <C extends Traversable<T>, T> C retainAll(C source, Iterable<? extends T> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        final Set<T> removed = HashSet.ofAll(elements);
+        return (C) source.filter(removed::contains);
     }
 
-    static <T, U, R extends Traversable<U>> R scanRight(Traversable<? extends T> traversable,
-                                                        U zero, BiFunction<? super T, ? super U, ? extends U> operation, Function<Iterator<U>, R> finisher) {
+    static <T> Iterator<T> reverseIterator(Iterable<T> iterable) {
+        if (iterable instanceof java.util.List) {
+            final java.util.List<T> list = (java.util.List<T>) iterable;
+            return new Iterator<T>() {
+                final java.util.ListIterator<T> delegate = list.listIterator(list.size());
```

Comment:
\[Checkstyle\] ERROR: Variable 'delegate' must be private and have accessor methods\.